glean-core/src/upload/result.rs

```diff
@@ -64,3 +70,122 @@ impl From<u32> for UploadResult {
         }
     }
 }
+
+impl UploadResult {
+    fn get_metric_label(&self) -> Option<&str> {
+        match self {
+            UploadResult::HttpStatus(200..=299) => None,
+            UploadResult::HttpStatus(400..=499) => Some("status_code_4xx"),
+            UploadResult::HttpStatus(500..=599) => Some("status_code_5xx"),
+            UploadResult::HttpStatus(_) => Some("status_code_unknown"),
+            UploadResult::UnrecoverableFailure => Some("unrecoverable"),
+            UploadResult::RecoverableFailure => Some("recoverable"),
+        }
+    }
+
+    fn get_metric(&self) -> Option<CounterMetric> {
+        if let Some(label) = self.get_metric_label() {
```

Comment:
It probably doesn't matter in a practical sense, but to make this work more like how labeled metrics work in the language bindings, I don't think we should create the metric "from scratch" like this each time.  Instead, lets add an object for `glean.upload.ping_upload_failure` to `internal_metrics.rs` and then call `new_metric_with_name` on it to get the specific labeled metric that we want.

Reply:
Well, I made this changes and there is much less code to this PR now :D